src/vos/vos_common.c

```diff
@@ -325,7 +325,21 @@ vos_tls_fini(void *data)
 {
 	struct vos_tls *tls = data;
 
-	D_ASSERT(d_list_empty(&tls->vtl_gc_pools));
+	/* All GC callers should have exited, but they can still leave
+	 * uncleaned pools behind. It is OK to free these pool handles with
+	 * leftover, because GC can clean up leftover when it starts again.
+	 */
+	D_ASSERTF(tls->vtl_gc_running == 0, "GC running = %d\n",
+		  tls->vtl_gc_running);
+
+	while (!d_list_empty(&tls->vtl_gc_pools)) {
```

Comment:
you can check but may be a little cleaner to do



>    while ((pool = d_list_pop_entry(....)) != NULL)
>         gc_del_pool(pool);

    
I think it leaves the link initialized.

Reply:
I'm not a fan of using d_list_pop_entry() != NULL, it will generate very long line and I also have to define the temporary variable outside the scope of the loop.

Reply:
I'm fine with removing the assert like what we did in pool_tls_fini() for this moment, but I think eventually we still need to properly drop references of ds_cont_hdl/ds_cont_child/ds_pool_child by local cont/pool close/disconnect on normal server shutdown (DAOS-6623), then we'll be able to put back the asserts or error messages in the module TLS fini functions, to me, those strict checking on TLS fini is helpful on revealing real reference leak defects.

See vos_pool_close(), if everything is properly dropped on shutdown, the pool should have been removed from the GC list.

Reply:
My intention is leaving the pool on GC even it's closed, because it is possible user delete a huge container and disconnect from the pool, the space is not going to be freed and it might cause space issues.

Reply:
VOS pool is only closed on server shutdown or pool destroy. Disconnecting pool won't cause VOS pool close.